### ISSUE TITLE
Add networks parameter to docker driver

### ISSUE THREAD

**Issue by abtreece** [metacloud/molecule#650]
# Feature request
Add the networks parameter to the docker driver.

The default bridge network does not support automatic service discovery. I created another bridge network to enable this feature, but am unable to launch my molecule controlled container on this network.

# Molecule and Ansible details

```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
neutron:kong abtreece$ molecule --version
molecule, version 1.16.1
```

- Molecule installation method: pip
- Ansible installation method: pip

# Desired Behaviour

Would like to be able to specify a container network other than the default bridge. Adding the links parameter would be an additional bonus although it is a deprecated feature.

# Actual Behaviour (Bug report only)

N/A

**Comment by kireledan** [comment]
I'll be able to get to this next week when I'm on my winter break if no one else has done it by then.

**Comment by retr0h** [comment]
@kireledan would be great if you could tackle it.

**Comment by kireledan** [comment]
The docker driver rework I have will cover this. I'll make a PR for it in a couple days.

**Comment by retr0h** [comment]
@kireledan Any chance you were able to tackle this?

**Comment by kireledan** [comment]
Yeah - currently just trying to get the tests to pass because it throws an error code 1 even though it doesn't error out.

**Comment by retr0h** [comment]
FYI - I just fixed functional testing on master earlier today.  For some reason I ran into a bug I never ran into before.

**Comment by kireledan** [comment]
Roger. Ill merge that into my branch. Currently my error is just that in using ansible to run the driver, the warning that ansible pops up with about localhost causes the test to fail immediately.

**Comment by lhoss** [comment]
we also desperately require this feature 👍 
as it (using a custom network/bridge) seems to be the cleanest way to deploy a cluster of docker containers with *working* hostname resolution (for ex. required to test a multi-host hadoop/hdfs deployment in our case) without any custom /etc/hosts hacks.
ps: I couldn't find any branch/commit.

As I'm totally new to molecule, I then wonder where's the best place in a molecule build to create that network ( `docker network create --driver bridge isolated_nw` ) ?https://docs.docker.com/engine/userguide/networking/

**Comment by kireledan** [comment]
I currently have it pushed to my playbook_proto branch. It is currently functional and does support networks, but it does not pass the unit tests because of some strange issues I'm having at the moment. https://github.com/kireledan/molecule/tree/playbook_proto

**Comment by abtreece** [comment]
@kireledan what's the reason you've chosen to use Ansible's docker module rather than just continuing to use docker-py directly? It would seem like we would have better access to the docker api if didn't implement a middle man.

**Comment by kireledan** [comment]
The reasoning behind the usage of the docker module was to be able to provide support for a wider number of options. Also, less maintenance due to the fact that the ansible docker module would already be updated to support future changes to the docker-py library. Plus it's a bit cleaner in my opinion.

**Comment by retr0h** [comment]
@kireledan @abtreece for this pull request, we should not be changing the way docker works in Molecule.  We should continue to docker-py.  However, in Molecule v2.  We will be relying on Ansible to manage instance state.  However, that will be marked as a v2 so users can make appropriate changes to their roles.

@abtreece The future of molecule is to use Ansible as a middle man. We spend more time adding features that already exist upstream into Molecule than we should.  If you were using the Ansible modules to manage the container, you would have this functionality already, w/o waiting on us.  A bug report never would have been filed.

**Comment by abtreece** [comment]
Perhaps then @kireledan can continue working his branch toward V2 and i'll attempt to add some basic networking functionality in for 1.20 using the current implementation of docker-py.

How far away from V2 are we?

**Comment by retr0h** [comment]
> How far away from V2 are we?

It's a huge rewrite and unless I run into some major issue, 3-6 months.

http://molecule.readthedocs.io/en/v2/

**Comment by buggsdummy** [comment]
Will this attempt to fix DNS issues as well?

**Comment by retr0h** [comment]
> Will this attempt to fix DNS issues as well?

What DNS issue?

**Comment by retr0h** [comment]
@kireledan I'll very likely incorporate the work you did into Ansible v2.  Looking at your branch, it doesn't appear you are passing AnsibleExec any form of inventory.  I think we need to give the loader the path to inventory some how.

**Comment by kireledan** [comment]
Okay - Yeah I'll be sure to submit the PR there for the docker change. I'll try to incorporate the same change to go along with the vagrant ansible module as well.

**Comment by retr0h** [comment]
Implemented by #725.

**Comment by lhoss** [comment]
I'm almost certain @buggsdummy is referring to having dns/autodiscovery between the containers (on the same bridge), which unfort.  was not addressed here  
(also my bad, that I cheered to early, and only found out now)
Well I created a new issue to support this -> https://github.com/metacloud/molecule/issues/748

Till then will have to keep a cumbersome work-around (using dnsmasq etc), as you can see here:
https://github.com/teralytics/ansible-hdfs/blob/master/playbook.yml#L11
